### PR TITLE
Use `stable/cloud` when Automatic Upgrades is on.

### LINKED ISSUE #22731: Cloud customers should get stable/cloud releases in Teleport Discover
All Cloud tenants who sign up after we roll out automatic upgrades will need to be served install scripts that use `stable/cloud` channel to install Teleport.

To achieve that, let's introduce an environment variable that the Cloud team will set for all new tenants who will be enrolled in automatic upgrades:

```
TELEPORT_AUTOMATIC_UPGRADES=yes
```

When this variable is set on the auth server, let's make sure that:

- [ ] It is returned as part of the auth server's [PingResponse](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/client/proto/authservice.proto#L535-L554) (e.g. under `ServerFeatures.AutomaticUpgrades`).
- [ ] It is propagated to proxy's [PingResponse](https://github.com/gravitational/teleport/blob/master/api/client/webclient/webclient.go#L277-L290) as well.
- [ ] All agent install scripts in Teleport Discover flows use `stable/cloud` deb/rpm channel.
- [ ] Helm chart flow in Teleport Discover shows command that enables auto-upgrades as well.
  - the chart must have the following values
    ```yaml
    updater:
        enabled: true
        releaseChannel: "stable/cloud"
    highAvailability:
      replicaCount: 2
      podDisruptionBudget:
        enabled: true
        minAvailable: 1
     ```
   - the helm command line must set the version based on the auth version or a curl subcommand
     ```
     helm install --version 12.1.2
     # or
     helm install -- version "$([curl https://updates.gravitational.io/stable/cloud/version)"
     ```
   - the values MUST NOT contain `teleportVersionOverride` (and this is also true for non-automatic-upgrade clusters, we must use `--version` instead, else we'll end up with compatibility issues)
    

- [ ] EC2 auto-discovery script uses `stable/cloud` deb/rpm channel as well.

### PR DESCRIPTION
**Pull request by marcoandredinis**
Teleport provides scripts to install teleport agents/services. 
Those scripts use YUM/DEB repositories when possible. Each repo has multiple channels:
- stable/v11
- stable/v12
- cloud/stable

We want to ensure that if the cluster is running in the cloud and automatic upgrades is on (auth service was started with TELEPORT_AUTOMATIC_UPGRADES=yes teleport ...), then the installation script must offer the `cloud/stable` channel.

This PR changes the following scripts:
- Discover Install Node script
- Discover Install Database Service script
- Install App script
- EC2 default-installer and EC2 default-agentless-installer

Demo:
Ping response 
`curl --silent -k https://127.0.0.1.nip.io:3080/webapi/ping | jq`
```json
{
  "auth": {
    "type": "local",
    "second_factor": "off",
    "local": {
      "name": ""
    },
    "private_key_policy": "none",
    "has_motd": false
  },
  "proxy": {
    "kube": {
      "enabled": true,
      "listen_addr": "0.0.0.0:3080"
    },
    "ssh": {
      "listen_addr": "0.0.0.0:3080",
      "tunnel_listen_addr": "0.0.0.0:3080",
      "web_listen_addr": "0.0.0.0:3080",
      "public_addr": "127.0.0.1.nip.io:3080"
    },
    "db": {
      "postgres_listen_addr": "0.0.0.0:3080",
      "mysql_listen_addr": "0.0.0.0:3080"
    },
    "tls_routing_enabled": true
  },
  "server_version": "13.0.0-dev",
  "min_client_version": "12.0.0",
  "cluster_name": "lenix",
  "automatic_upgrades": true
}
```

Install Node script
image
Install DB script
image
EC2 installer script
image
Helm chart installation
image

This needs a counterpart PR in `teleport.e` which I'll open as soon as this one gets merged

Fixes https://github.com/gravitational/teleport/issues/22731

### CODE REVIEW

File: web/packages/teleport/src/services/ping/makePing.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+import { PingResponse } from './types';
+
+export default function makePing(json: any): PingResponse {
+  json = json || {};
+  const { automatic_upgrades = {} } = json;
+
+  return {
+    automaticUpgrades: automatic_upgrades === true,
```
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply web/packages/teleport/src/services/ping/makePing.ts:24]: `automatic_upgrades` may not be present in the backend response
If we remove the `=== true`, then `automaticUpgrades` resolves to `undefined` which then means extra checks when we need to use it

File: web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx [hunk 2]
```diff
@@ -303,19 +303,31 @@ const generateCmd = (data: {
...
 roles: kube
 authToken: ${data.tokenId}
 proxyAddr: ${data.proxyAddr}
 kubeClusterName: ${data.clusterName}
-teleportVersionOverride: ${data.clusterVersion}
-enterprise: ${data.isEnterprise}
 labels:
     teleport.internal/resource-id: ${data.resourceId}
-EOF
+${extraYAMLConfig}EOF
```
Comment (hugoShaka) [inline web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx:328]: can we enable replication and pod distruption budget for everyone? we would need it at least if `data.automaticUpgradesEnabled`. Would you prefer to do it in another PR?

Comment (marcoandredinis) [reply web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx:328]: I think we should do it in another PR.
Just because we should try to expose as less params as possible in those scripts to improve UX.

As for the change itself, if they depend then we should ensure that at the helm chart level, right?
There's also an option to set those values as default.

I can open a PR for that, but I'll need your review

Comment (hugoShaka) [reply web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx:328]: I don't think we want to enable HA by default on the chart, this would be a significant breaking change and according to @webvictim a non-negligible amount of users don't want nor need replication.

The chart could have a smart default system like "default to 2 when the updater is on, default to 1 if it is not, and follow the user input whatever happens," but it makes things less clear from a user perspective and is also absolutely terrible to implement in Helm.

We do need to advise cloud users to enable replication if they are enrolled in auto upgrades. I don't know about cloud users without auto upgrades.

For those reasons, I would address the HA in Teleport Discover for cloud users.

File: web/packages/teleport/src/services/ping/makePing.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/services/ping/makePing.ts:24]: That's fine, as you'll be doing either ```if (pingResponse.automaticUpgrades) {}
// or
if (!pingResponse.automaticUpgrades) {}```

Which evaluates it to truthy/falsey values

File: web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx (hunk 2 above)
Comment (marcoandredinis) [reply web/packages/teleport/src/Discover/Kubernetes/HelmChart/HelmChart.tsx:328]: I'll add HA like you suggested on the linked issue
```
highAvailability:
  replicaCount: 2
  podDisruptionBudget:
    enabled: true
    minAvailable: 1
```

> I don't know about cloud users without auto upgrades.

For auto upgrades to be on, it'll need to be a cloud cluster **and** the env var `TELEPORT_AUTOMATIC_UPGRADES` is set to a truthy value

File: web/packages/teleport/src/services/ping/makePing.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcoandredinis) [reply web/packages/teleport/src/services/ping/makePing.ts:24]: > Edit - can we make automatic_upgrades always available in the response?

We already have `automaticUpgrades`. Should I remove it in favor of `automatic_upgrades` or keep both?
It seems you are suggesting both, why would we need both?

[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/services/ping/makePing.ts]: As I understood it, `automatic_upgrades` is from the backend that we're converting to `automaticUpgrades` for the frontend

Regardless of what it is, it would be good to make sure the field is always in the backend response and set to `true` or` false`

[the remark this answers is not in the document]
Comment (marcoandredinis) [reply web/packages/teleport/src/services/ping/makePing.ts]: Got it :+1:

[the remark this answers is not in the document]
Comment (marcoandredinis) [reply web/packages/teleport/src/services/ping/makePing.ts]: After last [commit](https://github.com/gravitational/teleport/pull/23395/commits/5b8bc3bec13793f982656af88ab4ca9f58167f3e):

```
dinis@lenix ~> curl --silent -k https://127.0.0.1.nip.io:3080/webapi/ping | jq '.automatic_upgrades'
false
dinis@lenix ~> curl --silent -k https://127.0.0.1.nip.io:3080/webapi/ping | jq '.automatic_upgrades'
true
```

**Review by hugoShaka — COMMENTED** [review]
The Helm-related part LGTM

**Comment by marcoandredinis** [comment]
@lxea Can you please review the EC2 installers?

File: api/utils/bool.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ import (
 )
 
 // ParseBool parses string as boolean value,
-// returns error in case if value is not recognized
+// returns false, error in case if value is not recognized
 func ParseBool(value string) (bool, error) {
 	switch strings.ToLower(value) {
 	case "yes", "yeah", "y", "true", "1", "on":
```
Comment (r0mant) [inline api/utils/bool.go:30]: Why do we have `nope` but no `yup` I wonder 😄

File: lib/modules/modules.go [hunk 4]
```diff
@@ -162,11 +168,14 @@ func (p *defaultModules) PrintVersion() {
 
 // Features returns supported features
 func (p *defaultModules) Features() Features {
+	automaticUpgrades, _ := utils.ParseBool(os.Getenv(defaults.AutomaticUpgradesEnvar))
```
Comment (r0mant) [inline lib/modules/modules.go:171]: Tbh, instead of silently ignoring the error, I would handle the case with the unset env. var explicitly and log the error from ParseBool. This will help in troubleshooting if we accidentally set incorrect value for example.

File: lib/web/join_tokens.go [hunk 5]
```diff
@@ -52,6 +52,8 @@ import (
 const (
 	teleportOSSPackageName = "teleport"
 	teleportEntPackageName = "teleport-ent"
+
+	cloudStableChannelRepo = "cloud/stable"
```
Comment (r0mant) [inline lib/web/join_tokens.go:56]: The channel name is `stable/cloud`, not `cloud/stable`.
```suggestion
	cloudStableChannelRepo = "stable/cloud"
```

File: lib/web/join_tokens.go [hunk 6]
```diff
@@ -265,9 +274,12 @@ func (h *Handler) getAppJoinScriptHandle(w http.ResponseWriter, r *http.Request,
 func (h *Handler) getDatabaseJoinScriptHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params) (interface{}, error) {
 	scripts.SetScriptHeaders(w.Header())
 
+	useCloudStableChannelRepo := h.ClusterFeatures.AutomaticUpgrades && h.ClusterFeatures.Cloud
```
Comment (r0mant) [inline lib/web/join_tokens.go:277]: Why do we have `&& h.ClusterFeatures.Cloud` check here? Can we omit it, just checking the auto-upgrades flag should be enough? We want to make this work for self-hosted clusters as well.

**Review by r0mant — CHANGES_REQUESTED** [review]
@marcoandredinis Can you double-check the channel name everywhere please, it should be `stable/cloud`, not `cloud/stable`. And can you update it in the PR title and description as well.

File: lib/web/join_tokens.go (hunk 6 above)
Comment (marcoandredinis) [reply lib/web/join_tokens.go:277]: The channel name is `stable/cloud`
Do we want other non-cloud/self-hosted clusters to follow a `cloud` specific version?

It kind of makes sense for self-hosted to follow a "rolling" release channel, but calling it cloud may sound confusing, because they are not using the cloud offering.

Comment (marcoandredinis) [reply lib/web/join_tokens.go]: Agreed on not having self-hosted users using the `stable/cloud` channel (for helm, deb, yum, ...)

As for supporting self-hosted, it's not clear how we are going to do it
For a complete self-hosted management, users would need to set up the s3 bucket for helm charts, but also deb and yum repos, which seems a bit overkill

**Review by r0mant — APPROVED** [review]

File: lib/web/join_tokens.go (hunk 6 above)
Comment (r0mant) [reply lib/web/join_tokens.go]: @marcoandredinis @hugoShaka They will have to host their own version lock server but they won't need to host their own repos. They will just use ours. It's ok if we just tell them to use stable/cloud initially, as long as they control the version and are ok with running a version that follows cloud release schedule it should be fine.

**Review by ryanclark — APPROVED** [review]

**Review by lxea — APPROVED** [review]

File: lib/web/join_tokens.go (hunk 6 above)
Comment (r0mant) [reply lib/web/join_tokens.go]: Let's keep your implementation as-is for now.

**Comment by avatus** [comment]
Please include this fix in the backport: https://github.com/gravitational/teleport/pull/23727

File: api/utils/bool.go (hunk 3 above)
Comment (webvictim) [reply api/utils/bool.go]:
```diff
+ yeppers
```

In all seriousness though, this is a fairly well-trodden path in the Teleport codebase already... 😅 

https://github.com/gravitational/teleport/blob/5a65ba9e7323e15769c9e3abb5d3908c4abb8344/api/types/authentication.go#L894-L897

https://github.com/gravitational/teleport/blob/5a65ba9e7323e15769c9e3abb5d3908c4abb8344/api/utils/bool.go#L30-L33

Comment (webvictim) [reply api/utils/bool.go]: Also my personal favourite

https://github.com/gravitational/teleport/blob/5a65ba9e7323e15769c9e3abb5d3908c4abb8344/lib/config/configuration_test.go#L1433
